File: Python/valid-parentheses.py
from __future__ import print_function
# Time:  O(n)
# Space: O(n)

class Solution:
    # @return a boolean
    def isValid(self, s):
        stack, lookup = [], {"(": ")", "{": "}", "[": "]"}
        for parenthese in s:
            if parenthese in lookup:
                stack.append(parenthese)
            elif len(stack) == 0 or lookup[stack.pop()] != parenthese:
                return False
        return len(stack) == 0

if __name__ == "__main__":
    print(Solution().isValid("()[]{}"))
    print(Solution().isValid("()[{]}"))

File: C++/Valid parentheses.cpp
class Solution {
public:
    bool isValid(string s) {


        stack<char> st;
      for(int i=0;i<s.size();i++)
        {
            if(s[i]=='('|| s[i]=='{' || s[i]=='[')
                {
                st.push(s[i]);
            }
           else
               {
               if(st.size()==0)//closing 
                   {
                   return false;
                   }
               if ((st.top()=='(' && s[i]==')') || (st.top()=='{' && s[i]=='}') || ( st.top()=='[' && s[i]==']'))
                   {
                   
                   st.pop();
                   }
              
            
               else
                   {
                   return false;
                   }
               
            }
          }
        return(st.size()==0);
               
                
     
            
}
};
